runtime/serde/serde-cbor/common/src/aws/smithy/kotlin/runtime/serde/cbor/CborUtils.kt

```diff
@@ -0,0 +1,223 @@
+internal fun ByteArray.toULong() = foldIndexed(0uL) { i, acc, byte ->
+    acc or (byte.toUByte().toULong() shl ((size - 1 - i) * 8))
+}
+
+// Subtracts one from the given BigInteger
+internal fun BigInteger.minusOne(): BigInteger {
+    val digits = toString().toCharArray()
+    var index = digits.lastIndex
+
+    // Process the digits from right to left
+    while (index >= 0) {
+        if (digits[index] > '0') {
+            digits[index] = digits[index] - 1
+            break
+        } else {
+            digits[index] = '9'
+            index--
+        }
+    }
+
+    // Remove leading zeroes
+    val result = digits.concatToString().trimStart('0')
+
+    return if (result.isEmpty()) BigInteger("0") else BigInteger(result)
+}
+
+// Adds one to the given BigInteger
+internal fun BigInteger.plusOne(): BigInteger {
+    val digits = toString().toCharArray()
+    var index = digits.lastIndex
+
+    // Process the digits from right to left
+    while (index >= 0) {
+        if (digits[index] == '9') {
+            digits[index] = '0'
+            index--
+        } else {
+            digits[index] = digits[index] + 1
+            return BigInteger(digits.concatToString())
+        }
+    }
+
+    // If all digits were '9', prepend '1'
+    return BigInteger("1${digits.concatToString()}")
+}
+
+// Converts a [BigInteger] to a [ByteArray].
+internal fun BigInteger.asBytes(): ByteArray {
+    var decimal = this.toString().removePrefix("-")
+    val binary = StringBuilder()
+
+    // Convert decimal to binary
+    while (decimal != "0") {
+        val temp = StringBuilder()
+        var carry = 0
+        for (c in decimal) {
+            val num = carry * 10 + c.digitToInt()
+            temp.append(num / 2)
+            carry = num % 2
+        }
+        binary.insert(0, carry)
+
+        decimal = temp
+            .dropWhile { it == '0' }
+            .ifEmpty { "0" }
+            .toString()
+    }
+
+    // Ensure the binary string length is a multiple of 8
+    val zeroPrefixLength = (8 - binary.length % 8) % 8
+    val paddedBinary = "0".repeat(zeroPrefixLength) + binary
+
+    // Convert each set of 8 bits to a byte
+    return paddedBinary.chunked(8)
+        .map { it.toUByte(radix = 2).toByte() }
+        .toByteArray()
+}
+
+// Converts a [ByteArray] to a [String] representing a BigInteger.
+internal fun ByteArray.toBigInteger(): BigInteger {
+    var decimal = "0"
+
+    // Iterate through each byte in the array
+    for (byte in this) {
+        val binaryString = byte.toUByte().toString(2).padStart(8, '0') // Convert each byte to an 8-bit binary string
+
+        // For each bit, update the decimal string
+        for (bit in binaryString) {
+            decimal = decimal.multiplyByTwo() // Multiply current decimal by 2 (shift left)
+            if (bit == '1') {
+                decimal = decimal.addOne() // Add 1 if the bit is 1
+            }
+        }
+    }
+
+    return BigInteger(decimal)
+}
```

Comment:
**Suggestion**: Similar comment as above—let's codify these in our KMP `BigInteger` and delegate to JVM's `BigInteger.toByteArray()` and ``BigInteger(`val`: ByteArray)``.